### ISSUE TITLE
AutoGluon Tabular ensembling fails with NaN-related error

### ISSUE THREAD

**Issue by ecoskian** [awslabs/autogluon#932]
Hello,

Just tried using AutoGluon on a Pandas DataFrame with the following:

```
task.fit(train_data=X_train, label='y', output_directory=models_dir, time_limits=3600, excluded_model_types=['KNN'])
```

None of the columns in the training data contains NaNs, however the first round of ensembling fails as follows.

Any help appreciated.

```
Fitting model: weighted_ensemble_k0_l1 ... Training model for up to 360.0s of the 2650.99s of remaining time.
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 856, in _train_and_save
    model = self._train_single(X_train, y_train, model, X_val, y_val, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 825, in _train_single
    model.fit(X=X_train, y=y_train, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 233, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/weighted_ensemble_model.py", line 23, in _fit
    super()._fit(X, y, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, compute_base_preds=compute_base_preds, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/stacker_ensemble_model.py", line 129, in _fit
    super()._fit(X=X, y=y, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/bagged_ensemble_model.py", line 127, in _fit
    model_base.fit(X_train=X, y_train=y, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 233, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/greedy_weighted_ensemble_model.py", line 37, in _fit
    self.model = self.model.fit(X_train, y_train, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/tuning/ensemble_selection.py", line 48, in fit
    self._fit(predictions=predictions, labels=labels, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/tuning/ensemble_selection.py", line 98, in _fit
    scores[j] = self.metric._optimum - calculate_score(
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/metrics/__init__.py", line 349, in calculate_score
    cprediction = sanitize_array(prediction)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/metrics/util.py", line 11, in sanitize_array
    maxi = np.nanmax(a[np.isfinite(a)])
  File "<__array_function__ internals>", line 5, in nanmax
  File "/home/eco/anaconda3/lib/python3.8/site-packages/numpy/lib/nanfunctions.py", line 434, in nanmax
    res = np.fmax.reduce(a, axis=axis, out=out, **kwargs)
Warning: Exception caused weighted_ensemble_k0_l1 to fail during training... Skipping this model.
Traceback (most recent call last):
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 856, in _train_and_save
    model = self._train_single(X_train, y_train, model, X_val, y_val, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 825, in _train_single
    model.fit(X=X_train, y=y_train, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 233, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/weighted_ensemble_model.py", line 23, in _fit
    super()._fit(X, y, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, compute_base_preds=compute_base_preds, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/stacker_ensemble_model.py", line 129, in _fit
    super()._fit(X=X, y=y, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/bagged_ensemble_model.py", line 127, in _fit
    model_base.fit(X_train=X, y_train=y, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 233, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/models/ensemble/greedy_weighted_ensemble_model.py", line 37, in _fit
    self.model = self.model.fit(X_train, y_train, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/tuning/ensemble_selection.py", line 48, in fit
    self._fit(predictions=predictions, labels=labels, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/ml/tuning/ensemble_selection.py", line 98, in _fit
    scores[j] = self.metric._optimum - calculate_score(
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/metrics/__init__.py", line 349, in calculate_score
    cprediction = sanitize_array(prediction)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/utils/tabular/metrics/util.py", line 11, in sanitize_array
    maxi = np.nanmax(a[np.isfinite(a)])
  File "<__array_function__ internals>", line 5, in nanmax
  File "/home/eco/anaconda3/lib/python3.8/site-packages/numpy/lib/nanfunctions.py", line 434, in nanmax
    res = np.fmax.reduce(a, axis=axis, out=out, **kwargs)
ValueError: zero-size array to reduction operation fmax which has no identity
```

**Comment by Innixma** [comment]
Could you try the pre-release version of AutoGluon? (Ensure fresh venv is used): https://auto.gluon.ai/dev/install.html

Unfortunately with the error you provided, it is hard to reproduce, so if pre-release does not work you will want to provide the full training logs instead of just the error.

**Comment by ecoskian** [comment]
Thanks. Tried again using the pre-release but no luck. Full log below.

```
Beginning AutoGluon training ... Time limit = 3600s
AutoGluon will save models to "/home/eco/ag_models/"
AutoGluon Version:  0.0.16b20210206
Train Data Rows:    773540
Train Data Columns: 48
Preprocessing data ...
Using Feature Generators to preprocess the data ...
Fitting AutoMLPipelineFeatureGenerator...
	Available Memory:                    887030.47 MB
	Train Data (Original)  Memory Usage: 286.21 MB (0.0% of available memory)
	Inferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.
	Stage 1 Generators:
		Fitting AsTypeFeatureGenerator...
	Stage 2 Generators:
		Fitting FillNaFeatureGenerator...
	Stage 3 Generators:
		Fitting IdentityFeatureGenerator...
		Fitting CategoryFeatureGenerator...
			Fitting CategoryMemoryMinimizeFeatureGenerator...
	Stage 4 Generators:
		Fitting DropUniqueFeatureGenerator...
	Useless Original Features (Count: 2): ['feature_0', 'feature_1']
		These features carry no predictive signal and should be manually investigated.
		This is typically a feature which has the same value for all rows.
		These features do not need to be present at inference time.
	Types of features in original data (raw dtype, special dtypes):
		('bool', [])  :  1 | ['feature_2']
		('float', []) : 44 | ['feature_3', 'feature_4', 'feature_5', 'feature_6', 'feature_7', ...]
		('int', [])   :  1 | ['hour']
	Types of features in processed data (raw dtype, special dtypes):
		('category', []) :  1 | ['feature_2']
		('float', [])    : 44 | ['feature_3', 'feature_4', 'feature_5', 'feature_6', 'feature_7', ...]
		('int', [])      :  1 | ['hour']
	3.1s = Fit runtime
	46 features in original data used to generate 46 features in processed data.
	Train Data (Processed) Memory Usage: 279.25 MB (0.0% of available memory)
Data preprocessing and feature engineering runtime = 3.56s ...
AutoGluon will gauge predictive performance using evaluation metric: 'root_mean_squared_error'
	To change this, specify the eval_metric argument of fit()
Automatically generating train/validation split with holdout_frac=0.01, Train Rows: 765804, Val Rows: 7736
Excluded Model Types: ['KNN']
	Found 'KNN' model in hyperparameters, but 'KNN' is present in `excluded_model_types` and will be removed.
	Found 'KNN' model in hyperparameters, but 'KNN' is present in `excluded_model_types` and will be removed.
Fitting model: RandomForestMSE ... Training model for up to 3596.44s of the 3596.43s of remaining time.
	-0.0937	 = Validation root_mean_squared_error score
	258.67s	 = Training runtime
	0.15s	 = Validation runtime
Fitting model: ExtraTreesMSE ... Training model for up to 3303.96s of the 3303.96s of remaining time.
	-0.0939	 = Validation root_mean_squared_error score
	118.36s	 = Training runtime
	0.13s	 = Validation runtime
Fitting model: LightGBM ... Training model for up to 3143.25s of the 3143.24s of remaining time.
	-0.0933	 = Validation root_mean_squared_error score
	4.08s	 = Training runtime
	0.01s	 = Validation runtime
Fitting model: LightGBMXT ... Training model for up to 3137.93s of the 3137.92s of remaining time.
	-0.0929	 = Validation root_mean_squared_error score
	8.67s	 = Training runtime
	0.01s	 = Validation runtime
Fitting model: CatBoost ... Training model for up to 3129.12s of the 3129.11s of remaining time.
Warning: Sign differs between AG metric and CatBoost metric variants: root_mean_squared_error, flipping signs.
	-0.0932	 = Validation root_mean_squared_error score
	31.57s	 = Training runtime
	0.02s	 = Validation runtime
Fitting model: XGBoost ... Training model for up to 3097.52s of the 3097.52s of remaining time.
	-0.093	 = Validation root_mean_squared_error score
	67.42s	 = Training runtime
	0.1s	 = Validation runtime
Fitting model: NeuralNetMXNet ... Training model for up to 3029.99s of the 3029.99s of remaining time.
	nan	 = Validation root_mean_squared_error score
	322.72s	 = Training runtime
	0.28s	 = Validation runtime
Fitting model: NeuralNetFastAI ... Training model for up to 2706.99s of the 2706.98s of remaining time.
/home/eco/anaconda3/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
█
	Ran out of time, stopping training early.
█
	-0.092	 = Validation root_mean_squared_error score
	2789.87s	 = Training runtime
	26.72s	 = Validation runtime
█
Fitting model: WeightedEnsemble_L1 ... Training model for up to 360.0s of the -149.84s of remaining time.
	Warning: Exception caused WeightedEnsemble_L1 to fail during training... Skipping this model.
Traceback (most recent call last):
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/tabular/trainer/abstract_trainer.py", line 895, in _train_and_save
    model = self._train_single(X_train, y_train, model, X_val, y_val, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/tabular/trainer/abstract_trainer.py", line 865, in _train_single
    model.fit(X_train=X_train, y_train=y_train, **model_fit_kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/abstract/abstract_model.py", line 310, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/ensemble/weighted_ensemble_model.py", line 26, in _fit
    super()._fit(X_train, y_train, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, compute_base_preds=compute_base_preds, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/ensemble/stacker_ensemble_model.py", line 137, in _fit
    super()._fit(X_train=X_train, y_train=y_train, k_fold=k_fold, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/ensemble/bagged_ensemble_model.py", line 141, in _fit
    model_base.fit(X_train=X_train, y_train=y_train, time_limit=time_limit, **kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/abstract/abstract_model.py", line 310, in fit
    self._fit(**kwargs)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/greedy_ensemble/greedy_weighted_ensemble_model.py", line 41, in _fit
    self.model = self.model.fit(X_train, y_train, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/greedy_ensemble/ensemble_selection.py", line 50, in fit
    self._fit(predictions=predictions, labels=labels, time_limit=time_limit)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/models/greedy_ensemble/ensemble_selection.py", line 100, in _fit
    scores[j] = self.metric._optimum - calculate_score(
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/metrics/__init__.py", line 487, in calculate_score
    cprediction = sanitize_array(prediction)
  File "/home/eco/anaconda3/lib/python3.8/site-packages/autogluon/core/metrics/util.py", line 11, in sanitize_array
    maxi = np.nanmax(a[np.isfinite(a)])
  File "<__array_function__ internals>", line 5, in nanmax
  File "/home/eco/anaconda3/lib/python3.8/site-packages/numpy/lib/nanfunctions.py", line 434, in nanmax
    res = np.fmax.reduce(a, axis=axis, out=out, **kwargs)
ValueError: zero-size array to reduction operation fmax which has no identity
zero-size array to reduction operation fmax which has no identity
```

**Comment by Innixma** [comment]
Seems the issue is coming from:

```
Fitting model: NeuralNetMXNet ... Training model for up to 3029.99s of the 3029.99s of remaining time.
	nan	 = Validation root_mean_squared_error score
	322.72s	 = Training runtime
	0.28s	 = Validation runtime
```

Somehow, the Neural network is getting a NaN score... I've never seen this before and it isn't obviously reproducible on any dataset that I have. Furthermore, I can't even figure out a way to artificially get a NaN score output from `autogluon.core.metrics.root_mean_squared_error` which is the metric used in your run.

I would be very interested if you could re-run this with high verbosity and limit to only using the neural network model via
```
TabularPredictor(label).fit(..., hyperparameters={'NN': {}}, verbosity=4)
```
This would give the per-epoch val_score of the NN, and would help identify how it got a score of NaN.

**Comment by Innixma** [comment]
I've merged a PR which will not save any model that gets a NaN score, to avoid crashes in future: #937

However, we are still investigating the root cause of the issue. It seems to be triggered when a prediction is NaN by the model, and we have only been able to reproduce it one time across 100s of runs.

**Comment by Innixma** [comment]
This has been fixed in #954.
